src/util/file_watch.c

```diff
@@ -0,0 +1,331 @@
+struct file_watch_ctx {
+    struct tevent_context *ev;
+    const char *filename;
+    bool use_inotify;
+
+    struct config_file_inotify_check {
+        struct snotify_ctx *snctx;
+    } inotify_check;
+
+    struct config_file_poll_check {
+        TALLOC_CTX *parent_ctx;
+        struct tevent_timer *timer;
+        struct file_watch_callback *callbacks;
+    } poll_check;
+
+    fw_fn fn;
+    void *fn_arg;
+};
+
+
+static void poll_watched_file(struct tevent_context *ev,
+                             struct tevent_timer *te,
+                             struct timeval t, void *ptr);
+
+static errno_t create_poll_timer(struct file_watch_ctx *fw_ctx)
+{
+    struct timeval tv;
+
+    gettimeofday(&tv, NULL);
+    tv.tv_sec += FILE_WATCH_POLL_INTERVAL;
+    tv.tv_usec = 0;
+
+    fw_ctx->poll_check.timer = tevent_add_timer(fw_ctx->ev,
+                                                fw_ctx->poll_check.parent_ctx,
+                                                tv,
+                                                poll_watched_file,
+                                                fw_ctx);
+    if (!fw_ctx->poll_check.timer) {
+        talloc_free(fw_ctx);
+        return EIO;
+    }
+
+    return EOK;
+}
+
+static void poll_watched_file(struct tevent_context *ev,
+                              struct tevent_timer *te,
+                              struct timeval t, void *ptr)
+{
+    int ret, err;
+    struct stat file_stat;
+    struct file_watch_ctx *fw_ctx;
+    struct file_watch_callback *cb;
+
+    fw_ctx = talloc_get_type(ptr, struct file_watch_ctx);
+
+    for (cb = fw_ctx->poll_check.callbacks; cb; cb = cb->next) {
+        ret = stat(fw_ctx->filename, &file_stat);
+        if (ret < 0) {
+            err = errno;
+            DEBUG(SSSDBG_FATAL_FAILURE,
+                  "Could not stat file [%s]. Error [%d:%s]\n",
+                  fw_ctx->filename, err, strerror(err));
+            return;
+        }
+
+        if (file_stat.st_mtime != cb->modified) {
+            /* Parse the configuration file and signal the children */
+            /* Note: this will fire if the modification time changes into the past
+             * as well as the future.
+             */
+            DEBUG(SSSDBG_CRIT_FAILURE, "File [%s] changed\n", fw_ctx->filename);
+            cb->modified = file_stat.st_mtime;
+
+            /* Tell the monitor to signal the children */
+            fw_ctx->fn(fw_ctx->filename, fw_ctx->fn_arg);
+        }
+    }
+
+    ret = create_poll_timer(fw_ctx);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_FATAL_FAILURE,
+              "Error: File [%s] no longer monitored for changes!\n",
+              fw_ctx->filename);
+    }
+}
+
+
+static int watched_file_inotify_cb(const char *filename,
+                                  uint32_t flags,
+                                  void *pvt)
+{
+    struct file_watch_ctx *fw_ctx;
+
+    DEBUG(SSSDBG_TRACE_INTERNAL,
+          "Received inotify notification for %s\n", filename);
```

Comment:
Compare `filename` with `fw_ctx->filename` as a sanity check?

Reply:
Sure. I thought I had done it.